### PR TITLE
Add config parameter to change stop timeout during daemon shutdown

### PR DESCRIPTION
**Pull request by yongtang**
**\- What I did**

This fix tries to add a daemon config parameter `--shutdown-timeout` that specifies the timeout value to stop containers gracefully (before SIGKILL). The default value is 15s.

**\- How I did it**

The `--shutdown-timeout` parameter is added to daemon options and config file. It will also be updated during daemon reload.

**\- How to verify it**

Additional test cases have been added to cover the change.

**\- Description for the changelog**

Add `--shutdown-timeout` to `dockerd` to specify the timeout (in seconds) to stop containers gracefully before daemon exit.

**\- A picture of a cute animal (not mandatory but encouraged)**

This fix is related to #22471 and #22566.

Signed-off-by: Yong Tang yong.tang.github@outlook.com

### CODE REVIEW

**Comment by thaJeztah** [comment]
Ok, we're okay with adding this, so moving to code review.

can you rebase?

**Comment by yongtang** [comment]
Thanks @thaJeztah the pull request has been rebased.

**Comment by thaJeztah** [comment]
sorry @yongtang, needs another rebase

**Comment by yongtang** [comment]
Thanks @thaJeztah, the PR has been rebased.

**Comment by vdemeester** [comment]
LGTM 🐸

File: daemon/config.go
Comment (cpuguy83) [inline daemon/config.go]: `time.Duration` ?

File: integration-cli/docker_cli_daemon_test.go
Comment (cpuguy83) [inline integration-cli/docker_cli_daemon_test.go]:
```go
select {
  case <-s.d.wait:
  case <-time.After(5*time.Second):
}
```

File: integration-cli/docker_cli_daemon_test.go
Comment (cpuguy83) [inline integration-cli/docker_cli_daemon_test.go]: Same as above

File: integration-cli/docker_cli_daemon_test.go
Comment (yongtang) [inline integration-cli/docker_cli_daemon_test.go]: Thanks @cpuguy83. The PR has been updated.

File: integration-cli/docker_cli_daemon_test.go
Comment (yongtang) [inline integration-cli/docker_cli_daemon_test.go]: Thanks @cpuguy83. Done.

File: daemon/config.go
Comment (yongtang) [inline daemon/config.go]: @cpuguy83 One potential issue with `time.Duration` is that the default marshal/unmarshal of `time.Duration` is nano-seconds (cannot parse a string like "5s"). In other words, if user wants to specify 5 seconds, it will be `"shutdown-timeout": 5000000000`, which might not be ideal.

Maybe we could use a customized type, e.g., 
```
type Duration struct {
    time.Duration
}
func (d Duration) MarshalJSON() ([]byte, error) {
```
so that user could specify `5s` instead of `5000000000`?

Comment (cpuguy83) [inline daemon/config.go]: I would expect this to be handled on the client, not the API (even though I'm sure we probably do this elsewhere...)

Comment (yongtang) [inline daemon/config.go]: @cpuguy83 The `--shutdown-timeout` is the global timeout and is handled in `dockerd`. It could be unmarshaled from  `--config-file=/etc/docker/daemon.json`.

Another PR (#22566) is the per-container timeout which could be handled in client.

Maybe for this PR we use `seconds` , and for the other PR we could change the per-container timeout (in #22566) to `time.Duration`?

**Comment by vdemeester** [comment]
@yongtang needs a rebase :sweat:

**Comment by yongtang** [comment]
Thanks @vdemeester. The pull request has been rebased.

**Comment by thaJeztah** [comment]
ping @vdemeester LGTY?

**Comment by vdemeester** [comment]
LGTM 🐸

File: daemon/daemon.go
Comment (mlaventure) [inline daemon/daemon.go]: The wording feels to me like we are only going to start shutting down the containers after `ShutdownTimeout` seconds have expired.

Maybe use something like `start clean shutdown of all containers with a %d seconds timeout...`?

cc @thaJeztah for better wording :sweat_smile:

File: docs/reference/commandline/dockerd.md
Comment (mlaventure) [inline docs/reference/commandline/dockerd.md]: alignment issue

**Review by mlaventure — CHANGES_REQUESTED** [review]
Couple of nits.

LGTM otherwise.

Also, requires a rebase.

File: daemon/daemon.go
Comment (yongtang) [inline daemon/daemon.go]: Thanks @mlaventure @thaJeztah . I will temporarily settle with `start clean shutdown of all containers with a %d seconds timeout...` but let me know if there are other suggestions.

**Comment by yongtang** [comment]
The PR has been rebased and updated. Please let me know if there are other issues.

**Review by mlaventure — APPROVED** [review]

**Comment by mlaventure** [comment]
LGTM if still :green_heart: 

ping @vdemeester @cpuguy83 @thaJeztah

File: docs/reference/commandline/dockerd.md
Comment (thaJeztah) [inline docs/reference/commandline/dockerd.md]: don't forget to update the man page as well https://github.com/docker/docker/blob/master/man/dockerd.8.md

File: docs/reference/commandline/dockerd.md
Comment (thaJeztah) [inline docs/reference/commandline/dockerd.md]: It's only supported on Linux, or Windows as well? If so, the "Windows" example needs to be updated as well :smile:

**Comment by thaJeztah** [comment]
left some docs notes already; @mlaventure IIRC, systemd will kill all processes if the systemd (unit-file) timeout is exceeded; do you think we should mention that in the documentation?

Also, this'll need changes in the bash and zsh completion scripts @albers @sdurrheimer

**Comment by mlaventure** [comment]
@thaJeztah I wouldn't make it `systemd` specific. Most service manager would have a similar (configurable) behavior.

Mentioning that their service manager may kill docker regardless if the daemon is stopped via it wouldn't hurt though.

File: docs/reference/commandline/dockerd.md
Comment (yongtang) [inline docs/reference/commandline/dockerd.md]: Thanks @thaJeztah. The man page has been updated.

File: docs/reference/commandline/dockerd.md
Comment (yongtang) [inline docs/reference/commandline/dockerd.md]: Thanks @thaJeztah. The docs has been updated.

**Review by thaJeztah — APPROVED** [review]
LGTM

ping @vdemeester PTAL

**Review by vdemeester — APPROVED** [review]
docs LGTM 🐸

**Comment by vdemeester** [comment]
@yongtang needs a rebase though 👼

**Comment by yongtang** [comment]
Thanks @vdemeester. The PR has been rebased and all tests passed.

**Comment by mlaventure** [comment]
ping @thaJeztah @vdemeester final look before merging? :)

**Comment by vdemeester** [comment]
Still LGTM 👼
merging :stuck_out_tongue_closed_eyes:

**Comment by yongtang** [comment]
Thanks all for the review and help! 👍

**Comment by albers** [comment]
Added bash completion: #27550

**Pull request merged**
